src/Lexer.js

```diff
@@ -230,7 +243,15 @@ module.exports = class Lexer {
       }
 
       // top-level paragraph
-      if (top && (token = this.tokenizer.paragraph(src))) {
+      // prevent paragraph consuming extensions by clipping 'src' to extension start
```

Comment:
The only problem with blocking the paragraph tokenizer for block tokens this way is that the start will have to be the actual start of a different block element otherwise it will break a single paragraph into multiple paragraphs.

Blocking inlineText for inline tokens this way doesn't really matter because multiple inline text tokens render the same as a single inline text token.

For Example:

We want to create an underline block token when a line is surrounded by `:` characters

The potential start is `/^:/`

```markdown
first line
:underlined line:
last line
```

```html
<p>first line</p>
<u>underlined line</u>
<p>last line</p>
```

The problem is when it is given a potential start that is not an actual start

```markdown
first line
:no underlined line
last line
```

```html
<p>first line</p>
<p>:no underlined line
last line</p>
```

where it should output

```
<p>first line
:no underlined line
last line</p>
```

Reply:
Hm good catch. Can we just use the full token regex then instead of a start character? That's how `rules.js` does it. We have the paragraph regex search for an entire table, fence, etc, not just the start characters.

Reply:
Alternatively, we can make the `Paragraph` step in the Lexer behave the same as `inlineText` where adjacent tokens are merged:

```
lastToken = tokens[tokens.length - 1];
        if (lastToken && lastToken.type === 'text') {
          lastToken.raw += token.raw;
          lastToken.text += token.text;
        } else {
          tokens.push(token);
        }
```

This solves the problem for me except it causes spec test `new smartypants_code` to fail, which actually might not be a correct test. That test shows two adjacent lines of text rendering as separate paragraphs for some reason:

[new smartypants_code odd behavior](https://marked.js.org/demo/?text=%23%23%23%23%23%20The%20lexer%20shows%20the%20following%20text%20as%20two%20paragraphs%3B%20The%20text%20%60script%60%20in%20%60%3Cscript-custom%3E%60%20is%20throwing%20off%20the%20parser.%0A%0A%0AEnsure%20that%20text%20such%20as%20custom%20tags%20that%20happen%20to%0Abegin%20with%20the%20same%20letters%20as%20the%20above%20tags%20don%27t%0Amatch%20and%20thus%20benefit%20from%20Smartypants-ing.%0A%3Cscript-custom%3E--foo%3C%2Fscript-custom%3E%0A%60--foo%60%20%3Ccodebar%20--foo%20codebar%3E%0A%0A%0A%23%23%23%23%23%20Change%20it%20to%20%60scrit%60%20and%20we%20get%20a%20single%20paragraph%2C%20which%20should%20be%20the%20correct%20behavior.%0A%0AEnsure%20that%20text%20such%20as%20custom%20tags%20that%20happen%20to%0Abegin%20with%20the%20same%20letters%20as%20the%20above%20tags%20don%27t%0Amatch%20and%20thus%20benefit%20from%20Smartypants-ing.%0A%3Cscrit-custom%3E--foo%3C%2Fscript-custom%3E%0A%60--foo%60%20%3Ccodebar%20--foo%20codebar%3E%0A&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20true%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master)

Reply:
I would think requiring a blank link before a block token extension should be good enough for now.

Reply:
I don't know... It's such a common thing in the spec to be able to interrupt paragraphs. I feel like it's a pretty important feature, and with the solution above it's not hard to fix (and even corrects broken behavior in that unit test).

Reply:
html with the `sanitize` option currently creates a paragraph token instead of html so it would break

```markdown
test
<div></div>
```

[demo](https://marked.js.org/demo/?text=test%0A%3Cdiv%3E%3C%2Fdiv%3E&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20true%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20true%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master)

Reply:
> It's such a common thing in the spec to be able to interrupt paragraphs.

This is one of those areas where the CommonMark spec uses the spec to describe how CommonMark works not how it should work.

According to the original markdown [spec](https://daringfireball.net/projects/markdown/basics):

> A paragraph is simply one or more consecutive lines of text, separated by one or more blank lines. (A blank line is any line that looks like a blank line — a line containing nothing but spaces or tabs is considered blank.)

Reply:
> CommonMark spec uses the spec to describe how CommonMark works not how it should work.

Right, I get that. But for better or for worse, being able to interrupt paragraphs is the behavior a lot of users are probably used to and expect because Commonmark has taken over as more or less the standard. Even Github continued the tradition when they introduced tables that also interrupt paragraphs

So the question is do we let users create "Commonmark" extensions, or do we restrict them to "original markdown" style?

(I personally really want to make some extensions that use this style, so I'm biased...)

In any case, I tweaked the `paragraph` clipping method so now it only merges paragraphs if the previous one was affected by the clipping-to-'start' process. It works with your sanitize tag example now.  Take a look....? 😅

Reply:
That still wouldn't help if `<` was the potential start of an extension. It would think that `test` was clipped and combine it with `<div></div>`.

Any extensions are, by definition, not "CommonMark" extensions because they are not in the CommonMark spec. GitHubs extensions would be considered "GFM" extensions. I don't see why any block extension can't have a blank line above it. Can you give me some examples of extensions that wouldn't work if we required a blank line?

Reply:
I think the only way to make it work with my sanitize tag example would be to add a `space` token before the html `paragraph` token if the last token was a `paragraph`.

I'm not sure if that would be a breaking change. I doubt anyone relies on there not being a `space` token there.

Reply:
or that could be another example of an error that would be fixed by combining `paragraph` tokens? There is no spec for sanitized html in markdown.

Reply:
> ...examples of extensions that wouldn't work if we required a blank line?

Say a user wants to make a new type of `table` that uses a different syntax than a normal GFM table. So they insert a custom `fancyTables` Tokenizer. And it *works*, until they start replacing their old tables with the custom Markdown, and they realize that half of them don't work. Because a blank line is missing. "But GFM tables don't need a blank line...? What is going on here?"

It's not that the extension "doesn't work", but that for the user it doesn't work "like the rest of my Markdown". In my own use case, for example, my business documents already follow a particular format that takes advantage of being able to place paragraphs directly above or below tables, blocks of code, etc. as captions or introductory text, which for us is very handy for legibility when skimming through the source (Markdown is designed to be somewhat human readable on its own). If adding extensions means abandoning our existing template, it's a bit of a nuisance.

So yeah, not that the extension doesn't work, but that the formatting is broken. I hope that makes sense.

If it comes down to it, we can push without that feature, but I would still like to see it \*someday\*...?